### PR TITLE
Rewrite: Swift

### PR DESCRIPTION
**Pull request by zenangst**
To do:

- [x] Rewrite categories into extensions
       - [x] Rewrite Sync
       - [x] Rewrite NSManagedObject
       - [x] Rewrite NSEntityDescription
- [x] Migration of the Demos
- [x] Update README with Swift support
- [x] Update README with how to get iOS 7 support
- [ ] Fix failing test regarding custom primary key and custom mapping
- [ ] Fix failing test regarding custom primary key

### CODE REVIEW

**Comment by 3lvis** [comment]
Fantastic initiative Chris!!!! :scream:

**Comment by 3lvis** [comment]
This PR should contain migration of the Objective-C example too (in case we find some issues there).

Also, do we need a special `Tests-Swift` target? Maybe we can update the existing Tests target so it's also in Swift.

**Comment by zenangst** [comment]
@3lvis @RamonGilabert will be the first guy out the door. We will compile a list of things to do in the initial description so that anyone can pick up the PR.

File: Source/Sync.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+let SyncCustomPrimaryKey = "hyper.isPrimaryKey"
+let SyncCustomRemoteKey = "hyper.remoteKey"
+
+private extension NSEntityDescription {
+
+    func sync_localKey() {
+        var localKey: String?
+
+        for (key, attributedDescription) in self.propertiesByName {
+            if let userInfo: Dictionary = attributedDescription.userInfo {
```
Comment (kostiakoval) [inline Source/Sync.swift]: if let userInfo = attributedDescription.userInfo  - Skip type when possible
+ use 1 if let a = a, b = b, c  = c {
}

**Comment by RamonGilabert** [comment]
I added the categories, if you want to work on one and you cannot continue, then the other know where to start, anything else to add?

File: Source/Sync.swift (hunk 1 above)
Comment (zenangst) [inline Source/Sync.swift]: @kostiakoval agreed, I made this before updating Xcode yesterday.

**Comment by 3lvis** [comment]
A good way to approach this might be to make separate branches and send PRs to this one, in that way you can keep work in progress of the subtasks.

File: Source/Sync.swift (hunk 1 above)
Comment (3lvis) [inline Source/Sync.swift]: @kostiakoval @zenangst  `let SyncCustomPrimaryKey = "hyper.isPrimaryKey"`

Do we need to keep the `Sync` prefix here? Or we can drop that now

File: Source/Sync.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import UIKit
+import CoreData
+
+let SyncCustomPrimaryKey = "hyper.isPrimaryKey"
```
Comment (3lvis) [inline Source/Sync.swift]: @kostiakoval @zenangst  `let SyncCustomPrimaryKey = "hyper.isPrimaryKey"`

Do we need to keep the `Sync` prefix here? Or we can drop that now

**Comment by 3lvis** [comment]
@zenangst @RamonGilabert @kostiakoval: 

Would making Sync a struct work when using Objective-C? It doesn't need to be a subclass of NSObject

**Comment by zenangst** [comment]
@3lvis I think we could add an Objective-C demo project and find out :+1:

File: Source/Sync.swift (hunk 2 above)
Comment (kostiakoval) [inline Source/Sync.swift]: we can drop it. Swift adds hidden Module prefix.
you can access it by: 
`Sync.SyncCustomPrimaryKey` or just `SyncCustomPrimaryKey`
but it you need and extra namespace inside your module you can make a struct and put keys there 
```swift 
struct Keys {
let CustomPrimaryKey
...
}
```

**Comment by kostiakoval** [comment]
@3lvis no it wont. struct are not visible to swift.

**Comment by kostiakoval** [comment]
you have to mark it with @objc annotation or subclass NSObject.

here is list of no available stuff in objc from swift.
https://developer.apple.com/library/ios/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html#//apple_ref/doc/uid/TP40014216-CH10-XID_87

**Comment by RamonGilabert** [comment]
We said two spaces right?

File: Source/Sync.swift (hunk 1 above)
Comment (RamonGilabert) [inline Source/Sync.swift]: It says: Ambiguous use of userInfo

**Comment by zenangst** [comment]
@RamonGilabert correct! :)

**Comment by 3lvis** [comment]
Looks like Travis doesn't work with Swift 1.2

**Comment by 3lvis** [comment]
@kostiakoval What a piece of swift.

screen shot 2015-04-09 at 8 56 10 pm

**Comment by RamonGilabert** [comment]
Had to update the example to compile

**Comment by kostiakoval** [comment]
you can make class, protocol or function. This is what meant by word 'piece'

**Comment by 3lvis** [comment]
@kostiakoval I understood what you meant, I was having fun with what it says in the docs

**Comment by kostiakoval** [comment]
:D

**Comment by zenangst** [comment]
@3lvis fixed now

File: Source/Sync.swift [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+      if let property = property as? NSRelationshipDescription {
+        relationships.append(property)
+      }
+    }
+
+    return relationships
+  }
+
+  func sync_processRelationshipsUsingDictionary(objectDictionary dictionary: [NSObject : AnyObject], andParent parent: NSManagedObject?, dataStack: DATAStack!) {
+    let relationships = self.sync_relationships()
```
Comment (kostiakoval) [inline Source/Sync.swift]: Do we use `self.` for methods calls :scream:

File: Source/Sync.swift [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    return objects?.first as? NSManagedObject
+  }
+
+  public func process(#changes: [AnyObject],
+    inEntityNamed entityName: String,
+    dataStack: DATAStack,
+    completion: (error: NSError) -> Void) {
+      [self.process(changes: changes,
```
Comment (kostiakoval) [inline Source/Sync.swift]: What is those `[ ]` braces for here?

Comment (zenangst) [inline Source/Sync.swift]: Fixed

**Comment by kostiakoval** [comment]
:heart_eyes: @zenangst

**Comment by zenangst** [comment]
@kostiakoval do we need the objective-c annotation on all classes or just the one that's gonna be exposed to Objective-C?

**Comment by kostiakoval** [comment]
@zenangst  just the one that's gonna be exposed to Objective-C? 
and you can add prefix in them 
like @objc(HYP) Mapper -> would be HYPMapper in objc

File: Source/Sync.swift (hunk 3 above)
Comment (kostiakoval) [inline Source/Sync.swift]: @3lvis @zenangst how about this one?

**Comment by 3lvis** [comment]
WHY ARE THE TESTS STILL PASSING?!

File: Source/Sync.swift (hunk 3 above)
Comment (3lvis) [inline Source/Sync.swift]: @kostiakoval I don't know, the main difference with using `self` for properties is that you are forced to use `self.` in some cases, meanwhile I think `self.` for methods is completely optional.

**Comment by zenangst** [comment]
@3lvis because it's still running the Objective-C source? :sunglasses:

File: Source/Sync.swift (hunk 3 above)
Comment (3lvis) [inline Source/Sync.swift]: Do you guys have any use case where you are forced to do `self.methodName()` ?

Comment (kostiakoval) [inline Source/Sync.swift]: when capturing self in closure.
the same as for properties.

Comment (3lvis) [inline Source/Sync.swift]: Mind sharing an example?

Comment (kostiakoval) [inline Source/Sync.swift]:
```swift
class A {
  
  func testA(x: Int) -> Int {
    return x + 1
  }
  
  func other() {
    
    var arr = [1,2,3,4]
    arr.map {
      self.testA($0)
    }
  }
}
```

Comment (kostiakoval) [inline Source/Sync.swift]: must use `self.testA`

Comment (kostiakoval) [inline Source/Sync.swift]: Swift has the same problem with strong reference casting cycle. 
And not using self (everywhere except closures) help to spot it.

Comment (3lvis) [inline Source/Sync.swift]: Using self everywhere would mean that you don't even face this problem.

Comment (kostiakoval) [inline Source/Sync.swift]: Yes, exactly!! :heart_eyes: 

For using `self.` with properties I agree with your point (show that life-time is longer than a function)
but for function calls this don't apply

Comment (kostiakoval) [inline Source/Sync.swift]: > Using self everywhere would mean that you don't even face this problem 

what do you mean?
Using self everywhere don't solve the strong reference cycle problem.

File: Source/Sync.swift (hunk 4 above)
Comment (3lvis) [inline Source/Sync.swift]: HehehehehehehHehehehehehehHehehehehehehHehehehehehehHeheheheheheh

**Comment by RamonGilabert** [comment]
I'm having a problem here, the: `// Generated by Apple Swift version 1.2 (swiftlang-602.0.49.3 clang-clang-602.0.49)` file says that the class is duplicated. Also, I cannot call the public methods in Sync to add the values.

**Comment by RamonGilabert** [comment]
I would drop the: `inEntityName` and `inContext` in Swift, I would use `entityName:` and `context:`

**Comment by kostiakoval** [comment]
@RamonGilabert Yes!!  +1 from me

**Comment by zenangst** [comment]
We had some discussions at the office today about that (and other things).
@RamonGilabert could you share two examples? Before and after.

**Comment by RamonGilabert** [comment]
@zenangst we had:

```
class func process(#changes: [AnyObject],
    inEntityNamed: String,
    predicate: NSPredicate?,
    parent: NSManagedObject?,
    inContext: NSManagedObjectContext,
    dataStack: DATAStack,
    completion: (error: NSError) -> Void) {

  }
```

And my idea is to change the `inEntityNamed` and `inContext` to:

```
class func process(#changes: [AnyObject],
    entityName: String,
    predicate: NSPredicate?,
    parent: NSManagedObject?,
    context: NSManagedObjectContext,
    dataStack: DATAStack,
    completion: (error: NSError) -> Void) {

  }
```

**Comment by kostiakoval** [comment]
I agree with that example.
Other example I wouldn't rename is `fetchInContext:(...)` when it's name of the function

**Comment by zenangst** [comment]
I think it looks good, @3lvis, @markvaldy thoughts?

**Comment by markvaldy** [comment]
I like it too.

File: Source/Sync.swift [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+      }
+
+      return objects?.first as? NSManagedObject
+  }
+
+  public func process(#changes: [AnyObject],
+    entityName: String,
+    predicate: NSPredicate?,
+    dataStack: DATAStack,
+    completion: (error: NSError) -> Void) {
```
Comment (zenangst) [inline Source/Sync.swift]: Do we prefer `Void` over `()`? My vote is on `Void`, loads and loads of `()` can quickly become confusing.
What do you guys think?

Comment (markvaldy) [inline Source/Sync.swift]: +1 on ```Void```. I think it's more readable.

Comment (RamonGilabert) [inline Source/Sync.swift]: I would stick with Void too

**Comment by RamonGilabert** [comment]
So I'm stuck here, I cannot call Sync.process() in the example, I think the other stuff is good already.

**Comment by 3lvis** [comment]
Still not sold to the Sync.process stuff.

Current API reads much nicer 

```
[Sync changes:....
```

**Comment by kostiakoval** [comment]
don't use NSObject in Swift if you don't need it!
you can make it as an global function if you need

**Comment by zenangst** [comment]
@kostiakoval I meant the `Sync(changes:`, not having it live on `NSObject` :sunglasses:

File: Examples/AppNet/AppNet/ViewController.swift [hunk 6]
```diff
@@ -59,13 +60,13 @@ extension ViewController: UITableViewDataSource {
...
-    var cell = self.tableView.dequeueReusableCellWithIdentifier(Constanst.SYNCCellIdentifier) as UITableViewCell
+    var cell = self.tableView.dequeueReusableCellWithIdentifier(Constant.SYNCCellIdentifier) as! UITableViewCell
     let data = self.items[indexPath.row]
 
-    cell = UITableViewCell(style: UITableViewCellStyle.Subtitle, reuseIdentifier: Constanst.SYNCCellIdentifier)
+    cell = UITableViewCell(style: UITableViewCellStyle.Subtitle, reuseIdentifier: Constant.SYNCCellIdentifier)
 
     cell.textLabel?.text = data.text
-    cell.textLabel?.numberOfLines = 0
+    cell.textLabel?.numberOfLines = 1
```
Comment (3lvis) [inline Examples/AppNet/AppNet/ViewController.swift]: Was this needed?

File: Source/Sync.swift [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+import CoreData
+
+let CustomPrimaryKey = "hyper.isPrimaryKey"
+let CustomRemoteKey = "hyper.remoteKey"
+let DefaultLocalPrimaryKey = "remoteID"
+let DefaultRemotePrimaryKey = "id"
+
+private extension NSEntityDescription {
+
+  func sync_localKey() -> String {
```
Comment (3lvis) [inline Source/Sync.swift]: I think we can drop the `sync` now

File: Source/Sync.swift [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+    var localKey = DefaultLocalPrimaryKey
+
+    for (key, attributedDescription) in self.propertiesByName {
+      if let
```
Comment (3lvis) [inline Source/Sync.swift]: `if let` O_O

crazy

File: Source/Sync.swift [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        userInfo: Dictionary = attributedDescription.userInfo,
```
Comment (3lvis) [inline Source/Sync.swift]: Can this just be `userInfo = attributedDescription.userInfo,`?

File: Source/Sync.swift [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        customPrimaryKey = userInfo[CustomPrimaryKey] as? String
+        where customPrimaryKey == "YES" {
```
Comment (3lvis) [inline Source/Sync.swift]: Beautiful

File: Source/Sync.swift [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+          localKey = key as! String
+      }
+    }
+
+    return localKey
+  }
+
+  func sync_remoteKey() -> String {
```
Comment (3lvis) [inline Source/Sync.swift]: I think we can drop the prefix now 

`sync_`

File: Source/Sync.swift [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    return remoteKey
+  }
+
+}
+
+extension NSManagedObject {
+
+  public func sync_processRelationshipsUsingDictionary(objectDictionary dictionary: [NSObject : AnyObject],
```
Comment (3lvis) [inline Source/Sync.swift]: I think we can drop the prefix now 

`sync_`

File: Source/Sync.swift (hunk 9 above)
Comment (RamonGilabert) [inline Source/Sync.swift]: It says, Ambiguous use of userInfo...

Comment (3lvis) [inline Source/Sync.swift]: I guess it makes sense, it's declared as AnyObject

**Comment by zenangst** [comment]
Is dropping the prefix on Extensions a good idea? I guess it's the same debate as having them on Categories in Objective-C.

**Comment by 3lvis** [comment]
@zenangst I think they are not needed now. I remember Mr @kostiakoval posting out that things get prefixed behind the scenes.

`Sync.localKey`

Meanwhile in Objective-C land you'll have non-guaranteed results.

**Comment by zenangst** [comment]
I think that only applies to classes, extensions don't have namespacing as far as I know. Also, how does this affect Objective-C compatibility? Maybe it doesn't and then I'm all for it :)



—

**Comment by RamonGilabert** [comment]
Well, as the methods that use sync_ are used within the class, I think the compatibility would be fine!

**Comment by zenangst** [comment]
Right, as long as they are not public it should be fine. You still have the issue with potential collision with Core Framework classes, but for me that is not as critical.



—

File: Examples/AppNet/AppNet/ViewController.swift (hunk 6 above)
Comment (3lvis) [inline Examples/AppNet/AppNet/ViewController.swift]: Nevermind, I saw what happens with 0. Updated the sample code on develop

**Comment by markvaldy** [comment]
There could be potential collisions and if the system introduces a public method with the same name as ours, but as @zenangst said, I think it's not critical.

**Comment by 3lvis** [comment]
@markvaldy I thought that only applied to Objective-C.

Damn.

File: Source/Sync.swift [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+    var remote = DefaultRemotePrimaryKey
+    let local = localKey()
+
+    if local != DefaultLocalPrimaryKey {
+      remote = local.hyp_remoteString()
+    }
+
+    return remote
+  }
+
```
Comment (markvaldy) [inline Source/Sync.swift]: Redundant line?

Comment (zenangst) [inline Source/Sync.swift]: Fixed in 9716643

**Comment by zenangst** [comment]
@3lvis 
screen shot 2015-04-13 at 21 13 08

**Comment by RamonGilabert** [comment]
Oh man, I love that! :P

**Comment by 3lvis** [comment]
@zenangst Before we had "use this instead". Can we add this here? It's ok to see that it's deprecated but we should point out what's the new method or what has changed.

**Comment by zenangst** [comment]
@3lvis https://github.com/hyperoslo/Sync/commit/27be9ccf95891a340b2660d36addd4839dc70ea5

**Comment by 3lvis** [comment]
Damn, sucks big time that we'll have to change our minimum deployment target to iOS 8 when merging this.

File: Source/Sync.swift [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+    return remote
+  }
+}
+
+extension NSManagedObject {
+
+  public func processRelationshipsUsingDictionary(objectDictionary dictionary: [NSObject : AnyObject],
+    parent: NSManagedObject?,
+    dataStack: DATAStack!) {
```
Comment (markvaldy) [inline Source/Sync.swift]: Why do we use `DATAStack!` here?

Comment (3lvis) [inline Source/Sync.swift]: @markvaldy I don't if it has something do to with this but for what is worth `DATAStack` could be `nil` if your `modelName` is not found

Comment (markvaldy) [inline Source/Sync.swift]: @3lvis But `DATAStack!` is declared as Implicitly Unwrapped Optional only in this place, and in all others we have `DATAStack`.

Comment (3lvis) [inline Source/Sync.swift]: That could be a problem. We should have one test where DATAStack is nil, just in case.

Would you like to add that?

Comment (markvaldy) [inline Source/Sync.swift]: @3lvis I see that Chris has already changed to `DATAStack` in https://github.com/hyperoslo/Sync/commit/702a9db6c7f7e5f67f46ce7bc7f30cde66daa6d7
Trying to pass nil from Objective-C produces a warning "Null passed to a callee that requires a non-null argument". In Swift compiler will not let you do that. So I believe we are good there now.

**Comment by 3lvis** [comment]
Added a couple of things

screen shot 2015-04-14 at 12 05 28 am

**Comment by 3lvis** [comment]
Swift's way to formatting things doesn't help for our old Objective-C style where we would move things to newlines.

I get the feeling that we might be better off keeping everything in one line...

**Comment by 3lvis** [comment]
Like 

```swift
let entity = NSEntityDescription.entityForName(
  childEntityName,
  inManagedObjectContext: self.managedObjectContext!)
```

vs

```swift
let entity = NSEntityDescription.entityForName(childEntityName,
  inManagedObjectContext: self.managedObjectContext!)
```

vs

```swift
let entity = NSEntityDescription.entityForName(childEntityName, inManagedObjectContext: self.managedObjectContext!)
```

**Comment by 3lvis** [comment]
screen shot 2015-04-15 at 10 21 19 pm

**Comment by 3lvis** [comment]
screen shot 2015-04-15 at 10 23 33 pm

**Comment by 3lvis** [comment]
# US?

screen shot 2015-04-15 at 10 22 01 pm

**Comment by 3lvis** [comment]
Doesn't seem that there has been more activity here in a while and the codebase has changed quite a lot. Might be better to leave this one in Objective-C and start focus in Swift for future pods.

**Pull request closed without merging**